braindecode/datautil/windowers.py

```diff
@@ -60,9 +60,12 @@ def create_windows_from_events(
 
     list_of_windows_ds = []
     for ds in concat_ds.datasets:
+        if mapping is None:
+            # mapping event descriptions to integers from 0 on
+            mapping = {v: k for k, v in enumerate(
+                np.unique(ds.raw.annotations.description))}
 
-        # TODO: how to get events without 'stim' channel?
-        events = mne.find_events(ds.raw)
+        events, _ = mne.events_from_annotations(ds.raw, mapping)
```

Comment:
I think this is a good improvement over the stim channel-only method!

I believe there is still another use case that this approach doesn't cover though. What if you have annotations for some clinical events (e.g., sleep stages, seizures, etc.) that are not necessarily of the same length? For example, you might have a sleep stage annotation that is 10-min long, followed by another one that is only 60-s long, etc. Let's say you are interested in extracting as many non-overlapping 30-s windows as possible inside each annotation. Currently, this function would only let you extract windows inside a fixed region following each event found (specifically, between `trial_start_offset_samples` and `trial_stop_offset_samples`), that is the same across all annotations. In this case, you would then be stuck with only extracting windows out of the first 60 s of each sleep stage annotation.

A possible solution would be to use the `chunk_duration` argument of `events_from_annotations` which already covers this scenario.  The function `_compute_supercrop_inds` could then be repurposed to find the `i_trials`, `i_supercrop_in_trials`, `starts` and `stops` for each event spit out by `events_from_annotations`. Another possibility would be to reimplement the `chunk_duration` functionality into `_compute_supercrop_inds` if that's simpler for now. Eventually though we could see whether there would be interest on MNE's side to add these kinds of outputs to `events_from_annotations`.

Reply:
As far as I get it, `_compute_supercrop_inds` uses `trial_stop_offset_samples` to determine the length of an event. The events object we had earlier did not include any information about the event length, so we had to specify this.

The annotations object now includes a duration from which we can recover the `trial_stop_offset_samples` for each event, as specified by the experiment. We could set this to optional, if you want the whole event to be included, and rely on the durations from the annotations object. If `chunk_duration` is given, this will be our future window length and the `events_from_annotations` will return the right events.

Do you think that will work?

Reply:
Yes, I think that would work! The only thing I'd say is that `chunk_duration` is actually equivalent to `stride` rather than window `size` (unless we also use the `overlap` argument :P).

Reply:
True, it just gives you the onsets ;)

Anyway, If you try to get as many same length intervals into a trial, `chunk_duration` should be your `window size` as well as your `stride`, I guess.

There is one issue with mne though. When providing `chunk_duration` in `events_from_annotations`, the last samples are dropped, if the remaining part is smaller than `chunk_duration`. Our option whether to drop the last samples will become nasty to implement then.

Reply:
> Anyway, If you try to get as many same length intervals into a trial, `chunk_duration` should be your `window size` as well as your `stride`, I guess.

Isn't that the case only if you want an overlap of 0? 

> There is one issue with mne though. When providing `chunk_duration` in `events_from_annotations`, the last samples are dropped, if the remaining part is smaller than `chunk_duration`. Our option whether to drop the last samples will become nasty to implement then.

If you remember I think we started implementing something like this during the sprint, it was indeed not super clean. But in the end I think the logic is not too complicated: if drop_last = False, we find the onset of a window that ends exactly at the end of the annotation; if that offset is not already in the list of onsets, we add it.

Reply:
> Isn't that the case only if you want an overlap of 0?

True

> If you remember I think we started implementing something like this during the sprint, it was indeed not super clean. But in the end I think the logic is not too complicated: if drop_last = False, we find the onset of a window that ends exactly at the end of the annotation; if that offset is not already in the list of onsets, we add it.

Yes, we already had that. I didn't remember ;)

Reply:
As @robintibor mentioned, we discussed the chunk duration issue already a lot during the sprint. For now, we would like to leave that to our own function, as this is making it clear what happens.

I now changed the way, the event windower handles the `trial_offset_samples`, namely as a real offset from the annotation objects duration.